packages-exp/auth-exp/src/core/user/additional_user_info.test.ts

```diff
@@ -0,0 +1,152 @@
+
+      it('for valid JSON', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK, rawUserInfo}))!.profile).to.deep.eq(profile);
+      });
+
+      it('for missing JSON', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK}))!.profile).to.deep.eq({});
+      });
+    });
+
+    describe('determines new-user status', () => {
+
+      it('for new users by token response', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK, isNewUser: true}))!.isNewUser).to.be.true;
+      });
+
+      it('for new users by toolkit response kind', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK, kind: 'identitytoolkit#SignupNewUserResponse'}))!.isNewUser).to.be.true;
+      });
+
+      it('for old users', () => {
+        expect(fromIdTokenResponse(idTokenResponse({providerId: ProviderId.FACEBOOK}))!.isNewUser).to.be.false;
+      });
+    });
+
+    describe('creates generic AdditionalUserInfo', () => {
+
+      it('for custom auth', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.CUSTOM, rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.be.null;
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+
+      it('for anonymous auth', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({providerId: ProviderId.ANONYMOUS, rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.be.null;
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+      /*
+      it('for missing provider IDs in response but not in token', () => {
+        const additionalUserInfo = fromIdTokenResponse(idTokenResponse({rawUserInfo}))!;
+        const {isNewUser, providerId, username, profile} = additionalUserInfo;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.FACEBOOK);
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+      */
+    });
+
+    describe('returns null', () => {
+      it('for missing provider IDs', () => {
+        expect(fromIdTokenResponse(idTokenResponse({}))).to.be.null;
+      });
+    });
+  });
+});
+
+function idTokenResponse(partial: PartialIdTokenResponse): IdTokenResponse {
+  return {
+    idToken: 'Parsing logic not implemented',
+    refreshToken: 'Doesn\'t matter',
+    expiresIn: 'Doesn\'t matter',
+    localId: 'Doesn\'t matter',
+    kind: 'Not a new user response ',
+    ...partial
+  };
+}
+
+interface PartialIdTokenResponse {
+  providerId?: ProviderId;
+  isNewUser?: boolean;
+  rawUserInfo?: string;
+  screenName?: string | null;
```

Comment:
this becomes string | null | undefined, prefer just undefined  or null as the fallback